### PR TITLE
Port the governance contract to use `piecrust` version `0.1`

### PR DESCRIPTION
**Pull request by ureeves**
Ports the governance contract to use `piecrust` version `0.1` and its respective ABI.

The contract has a small change in its state structure to use `AtomicPtr<T>` for the `broker` and `authority`. These have to be set by the host at deployment time. If not set, contract functions involving them will panic.

Resolves: #854
See-also: #760

### CODE REVIEW

File: contracts/governance/Makefile [hunk 1]
```diff
@@ -3,18 +3,23 @@ help: ## Display this help screen
 		-E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | \
 		awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
 
+all: ## Build the stake contract
```
Comment (herr-seppia) [inline contracts/governance/Makefile:6]: `stake` leftover here

File: contracts/governance/src/state.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    ///
+    /// This function should only be called once, but it does support being
+    /// called multiple times.
+    pub fn set_authority(&self, authority: BlsPublicKey) {
+        let authority = Box::leak(Box::new(authority));
+        let last_authority = self
+            .authority
+            .swap(authority as *mut BlsPublicKey, Ordering::SeqCst);
+
+        // If the broker was already set we need to recoup the memory, otherwise
```
Comment (herr-seppia) [inline contracts/governance/src/state.rs:162]: the comment should be `if the authority [...]`

**Review by herr-seppia — APPROVED** [review]
LGTM with 2 small nits

File: contracts/governance/src/state.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        // we will leak memory.
+        if last_broker != ptr::null_mut() {
+            unsafe { Box::from_raw(last_broker) };
+        }
+    }
+
+    /// Set the authority in the contract.
+    ///
+    /// This function should only be called once, but it does support being
+    /// called multiple times.
```
Comment (miloszm) [inline contracts/governance/src/state.rs:155]: Nit - `idempotent` is the word I'd use here :)

**Review by miloszm — CHANGES_REQUESTED** [review]
Really nit, but hex::encode should go IMO

File: contracts/governance/src/state.rs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                panic!("Balance overflow");
+            }
+            *balance = new_balance.unwrap();
+        } else {
+            self.balances.insert(address, amount);
+        }
+    }
+
+    /// Subtract the value given from the specified address' balance.
+    /// If the address is not present nothing happens.
```
[the remark this answers is not in the document]
Comment (ureeves) [reply contracts/governance/src/state.rs:241]: The remainder is returned, not the original value, but I get the point

File: rusk-recovery/src/state.rs [hunk 5]
```diff
@@ -45,33 +45,27 @@ pub static FAUCET_KEY: Lazy<PublicSpendKey> = Lazy::new(|| {
...
-    let contract_id = governance.contract();
+    );
 
     let theme = Theme::default();
     info!(
         "{} {} governance to {}",
         theme.action("Deploying"),
         governance.name,
-        contract_id
+        hex::encode(module_id)
```
[the remark this answers is not in the document]
Comment (ureeves) [reply rusk-recovery/src/state.rs:61]: `ModuleId` does not implement `Display` in `piecrust-uplink` version `0.1.0`

**Review by miloszm — APPROVED** [review]
LGTM

**Comment by Daksh14** [comment]
LGTM, I was trying to run miri on the project but failed to catch any uninitiated var or UB since there's unsafe usage in the code. If we are sure about the safety of this PR, its a 👍 from me.a

**Pull request merged**